plugins/groups_manager.py

```diff
@@ -0,0 +1,312 @@
+            self.opts.load.insert(0, self.opts.merge)
+            self.opts.save.append(self.opts.merge)
+
+        # check that group is specified when editing is attempted
+        if (self.opts.description
+                or self.opts.display_order
+                or self.opts.translated_name
+                or self.opts.translated_description
+                or self.opts.user_visible is not None
+                or self.opts.packages):
+            if not self.opts.id and not self.opts.name:
+                raise dnf.cli.CliError(
+                    _("Can't edit group without specifying it (use --id or --name)"))
+
+    def load_input_files(self):
+        """
+        Loads all input xml files.
+        Returns True if at least one file was successfuly loaded
+        """
+        for file_name in self.opts.load:
+            file_comps = libcomps.Comps()
+            try:
+                if file_name.endswith('.gz'):
+                    # libcomps does not support gzipped files - decompress to temporary
+                    # location
+                    with gzip.open(file_name) as gz_file:
+                        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+                            try:
+                                shutil.copyfileobj(gz_file, temp_file)
+                                # close temp_file to ensure the content is flushed to disk
+                                temp_file.close()
+                                file_comps.fromxml_f(temp_file.name)
+                            finally:
+                                os.unlink(temp_file.name)
+                else:
+                    file_comps.fromxml_f(file_name)
+            except (IOError, OSError, libcomps.ParserError) as err:
+                # gzip module raises OSError on reading from malformed gz file
+                # get_last_errors() output often contains duplicit lines, remove them
+                seen = set()
+                for error in file_comps.get_last_errors():
+                    if error in seen:
+                        continue
+                    logger.error(error.strip())
+                    seen.add(error)
+                raise dnf.exceptions.Error(
+                    _("Can't load file \"{}\": {}").format(file_name, err))
+            else:
+                self.comps += file_comps
+
+    def save_output_files(self):
+        for file_name in self.opts.save:
+            try:
+                # xml_f returns a list of errors / log entries
+                errors = self.comps.xml_f(file_name, xml_options=COMPS_XML_OPTIONS)
+            except libcomps.XMLGenError as err:
+                errors = [err]
+            if errors:
+                if len(errors) > 1:
+                    errors.append(_("Error saving file {}").format(file_name))
+                for err in errors[:-1]:
+                    logger.error(err.strip())
+                raise dnf.exceptions.Error(errors[-1].strip())
```

Comment:
Feels like I'm missing something once again, but why append the "Error saving file" to the list of errors, only to then skip it when logging and pick it from the end of the list when throwing the exception?

If there's only one error, you'll raise it directly, but then you don't include the "Error saving file PATH" part... Seems like the error handling should be the same as when loading, which seems to make more sense?

Reply:
The problem here is that there is not one particular error that occurred during saving the file - there is whole list of them. So in case the particular error is not clear (i.e. number of errors returned is greater than one) I add a generic "Error saving file" message without any details (the details were logged before).

But there of course are other possibilities - for example log all errors except the last one and raise the last one as `dnf.exceptions.Error` hoping that it contains something reasonable. That actually might make sense given that I was not able to make libcomps emit more than one error during xml saving.

As far as error message wording - it sounds a bit clumsy (either because dnf.Error handler automatically adds 'Error: ' prefix or due to repeating the libcomps error). That is why I did not add the "Error saving file" part (but I'm not sure I can rely on libcomps adding the filename to its message):
```
Error: Error saving file "nnn/out.xml": Can't write file nnn/out.xml
Error: Can't save file "nnn/out.xml": Can't write file nnn/out.xml
Error: Can't write file nnn/out.xml
```

There is one more problem - libcomps writes some errors directly to stdout (sometimes even more than once) and do not return them in error array (but this is out of scope of this PR). So currently the full output in case of trying to save the xml to non-existend directory looks like this:
```
I/O error : No such file or directory
I/O error : No such file or directory
Error: Can't write file nonexistent/path/out.xml
```
The I/O error lines are written directly by libcomps and I cannot handle them.

Reply:
Okay, seems like you'd have to go fixing libcomps, which is probably not a priority now...

I think my point really is why not do it the same way as in `load_input_files()`, i.e. log all errors in the `errors` list and then raise an exception using the message from the exception you've caught. Now I'm not sure you've answered this question in your writeup :slightly_smiling_face:

Reply:
The thing is, that usually there is no exception caught. Just list of returned errors.

Reply:
Ok, I don't know, let's keep it this way then... can you add a short comment describing the situation so that future readers know what's up?